File: A5_ChatBot.py
import nltk
from nltk.chat.util import Chat,reflections
reflections = {
  "i am"       : "you are",
  "i was"      : "you were",
  "i"          : "you",
  "i'm"        : "you are",
  "i'd"        : "you would",
  "i've"       : "you have",
  "i'll"       : "you will",
  "my"         : "your",
  "you are"    : "I am",
  "you were"   : "I was",
  "you've"     : "I have",
  "you'll"     : "I will",
  "your"       : "my",
  "yours"      : "mine",
  "you"        : "me",
  "me"         : "you"
}
pairs = [
    [
        r"my name is (.*)",
        ["Hello %1, How are you today ?",]
    ],
    [
        r"hi|hey|hello",
        ["Hello", "Hey there",]
    ],
    [
        r"what is your name ?",
        ["I am a bot created by Yash Rajput. you can call me Wed!",]
    ],
	[
        r"what is the timing of hospital ?",
        ["Monday to Saturday - Morning 10am to Evening 6pm",]
    ],
	[
        r"what is the name of hospital ?",
        ["OPERA HSPITAL",]
    ],
	[
        r"what is the timing of hospital ?",
        ["Morning 10am to Evening 6pm",]
    ],
    [
        r"how are you ?",
        ["I'm doing good.How about You ?",]
    ],
    [
        r"sorry (.*)",
        ["Its alright","Its OK, never mind",]
    ],
    [
        r"I am fine",
        ["Great to hear that, How can I help you?",]
    ],
    [
        r"i'm (.*) doing good",
        ["Nice to hear that","How can I help you?:)",]
    ],
    [
        r"(.*) age?",
        ["I'm a computer program and Seriously you are asking me this?",]
    ],
    [
        r"what (.*) want ?",
        ["To help everyone",]
    ],
    [
        r"(.*) created ?",
        ["Yash created me using Python's NLTK library ","top secret :)",]
    ],
    [
        r"(.*) (location|city) ?",
        ['Vigyan Nagar, Kota',]
    ],
    [
        r"average salary of doctor",
        ["It's 10 lakhs"]
    ],
    [
        r"what is appointment fees per person?",
        ["100 Rs per person."]
    ],
    [
        r"Can a doctor come for home check-up?",
        ["Yes, the the emergency type situation only."]
    ],
    [
        r"(.*)raining in (.*)",
        ["No rain since last week here in %2","Damn its raining too much here in %2"]
    ],
    [
        r"how (.*) health(.*)",
        ["I'm a computer program, so I'm always healthy ",]
    ],
    [
        r"Are there any check-up packages for today?",
        ["Today free diabetes checkup is ongoing.",]
    ],
    [
        r"I have been having ear pain for the past two weeks. Which doctor should I consult.",
        ["It is better you consult with ENT specialist. Today Dr. (Ja,es Sebastian | George kurien) is available."]
    ],
    [
        r"i am looking for online guides and courses to learn data science, can you suggest?",
        ["Crazy_Tech has many great articles with each step explanation along with code, you can explore"]
    ],
    [
        r"quit",
        ["Bye take care. See you soon :) ","It was nice talking to you. See you soon :)"]
    ],
	[
		r"Why don\'t you (.*)",
        [
            "Do you really think I don't %1?",
            "Perhaps eventually I will %1.",
            "Do you really want me to %1?",
		]
	]
]
def chat():
	print("Please ask me a Question?")
	chat=Chat(pairs,reflections)
	chat.converse()

if __name__=="__main__":
	chat()

File: A5_Chatbot.cpp
#include <bits/stdc++.h>
using namespace std;
class questions {
    map<string, string> mp;
    int asked_que_ptr;
    vector<string> some_questions_to_be_asked;

public:
    questions() {
        asked_que_ptr = 0;
        mp["hi"] = "hello";
        mp["How are you"] = "I am fine, What about you?";
        mp["What's your name?"] = "My name is Wed. What's your name";
        mp["What's your age?"] = "I was created on march 2022 by Yash Rajput";
        mp["what's the timing of hospital?"] = "Morning 10am to Evening 6pm";
        some_questions_to_be_asked.push_back("What's your name?");
        some_questions_to_be_asked.push_back("What's your father's name?");
        some_questions_to_be_asked.push_back("What's your mother's name?");
        some_questions_to_be_asked.push_back("What's your problem?");
    }
    string accuracy(string question) {
        int accuracy = 0;
        string ans = "Sorry! unable to understand\n";
        map<char, int> mp_of_question;
        for (auto i : question) {
            mp_of_question[i]++;
        }
        for (auto i : mp) {
            int temp_accuracy = 0;
            map<char, int> map_of_question_in_dataset;
            for (auto k : i.first) {
                map_of_question_in_dataset[k]++;
            }
            for (char c = 'a'; c <= 'z'; c++) {
                temp_accuracy += min(map_of_question_in_dataset[c], mp_of_question[c]);
            }
            if (temp_accuracy > accuracy) {
                accuracy = temp_accuracy;
                ans = i.second;
            }
        }
        return ans;
    }
    void AskMeAnything() {
        cout << "Hey!!\n";
        cout << "What's your doubt\n";
        string question;
        cin.ignore();
        getline(cin, question);
        string ans = accuracy(question);
        cout << ans << '\n';
    }
    void addAnswer() {
        while (asked_que_ptr < some_questions_to_be_asked.size()) {
            cout << "If you want to stop please enter -1\n";
            cout << some_questions_to_be_asked[asked_que_ptr];
            string ans;
            cin.ignore();
            getline(cin, ans);
            if (ans == "-1") {
                break;
            }
            mp[some_questions_to_be_asked[asked_que_ptr]] = ans;
            asked_que_ptr++;
        }
    }
    void addManually() {
        cout << "Enter your Question\n";
        string que;
        cin.ignore();
        getline(cin, que);
        cout << "Enter your answer\n";
        cin >> mp[que];
    }
};
int main() {
    cout << "Welcome to OPERA HOSPITAL\n";
    cout << "Hi!! how can i help you?\n";
    questions obj;
    while (1) {
        cout << "\n1 Ask me a question\n";
        cout << "2 Want to give some answer\n";
        cout << "3 want to add some answer manually\n";
        cout << "0 Exit\n";
        int c;
        cin >> c;
        if (c == 0) {
            cout << "Hope you like our service if you are not satisfied than talk to our customer care '1800 256 452'\n";
            cout << "Thank you :)\n";
            break;
        } else if (c == 1) {
            obj.AskMeAnything();
        } else if (c == 2) {
            obj.addAnswer();
        } else if (c == 3) {
            obj.addManually();
        } else {
            cout << "Wrong choice :(\n";
            cout << "Please try again\n";
        }
    }
}